lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,214 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
-int pthread_mutex_trylock(pthread_mutex_t *m)
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+	.protocol = PTHREAD_PRIO_NONE,
+};
+
+static int mutex_recursive_take(pthread_mutex_t *m, int timeout)
+{
+	if (m->lock_count == 0) {
+		m->owner = pthread_self();
+	}
+
+	if (m->owner == pthread_self()) {
+		m->lock_count++;
+		if (m->lock_count > MUTEX_MAX_REC_LOCK) {
+			return EAGAIN;
+		}
+		return k_sem_take(m->sem, timeout);
+	}
+
+	return EPERM;
+}
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
 {
-	int key = irq_lock(), ret = EBUSY;
+	pthread_t this_thread;
+
+	this_thread = pthread_self();
+	int rc = 0;
+
+	if (m->owner == this_thread) {
+		if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+		rc = mutex_recursive_take(m, timeout);
+	} else {
+		m->lock_count = 1;
+		rc = k_sem_take(m->sem, timeout);
+	}
 
-	if (m->sem->count) {
-		m->sem->count = 0;
-		ret = 0;
+	if (rc == 0) {
+		m->owner = this_thread;
 	}
 
+	return rc;
+}
+
+/**
+ * @brief Lock POSIX mutex with non-blocking call.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_trylock(pthread_mutex_t *m)
+{
+	int key = irq_lock(), ret;
+
+	ret = acquire_mutex(m, K_NO_WAIT);
+
 	irq_unlock(key);
 
 	return ret;
 }
+
+/**
+ * @brief Intialize POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *attr)
+{
+	const pthread_mutexattr_t *mattr;
+
+	if (k_mem_pool_alloc(&posix_mutex_sem, &m->block,
+			sizeof(struct k_sem), 100) == 0) {
+		memset(m->block.data, 0, sizeof(struct k_sem));
+	}
+
+	m->sem = m->block.data;
+
+	m->owner = NULL;
+	m->lock_count = 0;
+
+	mattr = (attr == NULL) ? &def_attr : attr;
+
+	m->type = mattr->type;
+	m->protocol = mattr->protocol;
+
+	if ((m->type == PTHREAD_MUTEX_NORMAL) ||
+		(m->type == PTHREAD_MUTEX_ERRORCHECK)) {
+		k_sem_init(m->sem, 1, 1);
+	} else {
+		k_sem_init(m->sem, MUTEX_MAX_REC_LOCK, MUTEX_MAX_REC_LOCK);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Lock POSIX mutex with blocking call.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_lock(pthread_mutex_t *m)
+{
+	return acquire_mutex(m, K_FOREVER);
+}
+
+/**
+ * @brief Unlock POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_unlock(pthread_mutex_t *m)
+{
+	if (m->owner != pthread_self()) {
+		return EPERM;
+	}
+
+	if (m->lock_count == 0) {
+		return -1;
+	}
+
+	k_sem_give(m->sem);
+	m->lock_count--;
```

Comment:
This is racy. The decrement and comparison should happen while the semaphore is taken.

Reply:
http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutex_lock.html 

I maintained separate tracing for lock count. According to definition, pthread_mutex_lock increases count and while unlocking mutex it decrease count. That why I have used it separately.

Reply:
Still let me know if I am missing any of comments to be resolved